homeassistant/components/zha/__init__.py

```diff
@@ -369,6 +419,123 @@ def zdo_command(self, tsn, command_id, args):
+        if discovery_info['manufacturer'] and discovery_info['model'] is not \
+                None:
+            self._id = "{}.{}_{}_{}_{}{}".format(
+                self._domain,
+                slugify(discovery_info['manufacturer']),
+                slugify(discovery_info['model']),
+                self._ieeetail,
+                discovery_info['endpoint'].endpoint_id,
+                discovery_info.get('entity_suffix', '')
+            )
+        else:
+            self._id = "{}.zha_{}_{}{}".format(
+                self._domain,
+                self._ieeetail,
+                discovery_info['endpoint'].endpoint_id,
+                discovery_info.get('entity_suffix', '')
+            )
+
+
+class ZHASwitchEvent(ZHAEvent):
+    """Switch / remote event for zha"""
+
+    def __init__(self, hass, cluster, discovery_info):
+        """Initialize Switch."""
+        super().__init__(hass, cluster, SWITCH, discovery_info)
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0040):
+            self._hass.bus.fire(
+                OFF_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif command_id in (0x0001, 0x0041, 0x0042):
+            self._hass.bus.fire(
+                ON_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif command_id == 0x0002:
+            self._hass.bus.fire(
+                TOGGLE_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+
+
+class ZHALevelEvent(ZHAEvent):
+    """Switch / remote event for zha"""
+
+    def __init__(self, hass, cluster, discovery_info):
+        """Initialize Switch."""
+        super().__init__(hass, cluster, SWITCH, discovery_info)
+        self._level = 0
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0004):  # move_to_level, -with_on_off
+            self.set_level(args[0])
+        elif command_id in (0x0001, 0x0005):  # move, -with_on_off
+            # We should dim slowly -- for now, just step once
+            rate = args[1]
+            if args[0] == 0xff:
+                rate = 10  # Should read default move rate
+            self.move_level(-rate if args[0] else rate)
+        elif command_id == 0x0002:  # step
+            # Step (technically shouldn't change on/off)
+            self.move_level(-args[1] if args[0] else args[1])
+
+    def attribute_update(self, attrid, value):
+        """Handle attribute updates on this cluster."""
+        if attrid == 0:
+            self._level = value
+
+    def move_level(self, change):
+        """Increment the level."""
+        self.set_level(min(255, max(0, self._level + change)))
+
+    def set_level(self, level):
```

Comment:
This is called from a method annotated with `@callback`. This means that it is run _inside_ the event loop. Please also annotate this method with callback to indicate that. Preferably you prefix all methods like`async_set_level`, to indicate it's inside the event loop.

Reply:
cluster_command is a function called by zigpy. I may have added @callback incorrectly. Is there a way to determine if this is the case?

Reply:
I think that zigpy runs inside the event loop, check the internals of the module or look at how the other code that interacts with zigpy does it.